### PR TITLE
OpenShift API Conventions: no annotations convention

### PR DESCRIPTION
**Pull request by ingvagabund**
Based on https://github.com/openshift/enhancements/pull/1296/files#r1187548070

### CODE REVIEW

File: dev-guide/api-conventions.md [hunk 1]
```diff
@@ -428,6 +428,24 @@ The simplest line is "no functions".
 Functions can be added in a separate repo, possibly library-go if there are sufficient consumers.
 Helpers for accessing labels and annotations are not recommended.
 
+### No annotation-based APIs
+
+Do not use annotations for extending an API. Annotations may seem as a good candidate for introducing experimental/new
+API. Nevertheless, migration from annotations to formal schema usually never happens as it may lead to breaking changes
```
Comment (deads2k) [inline dev-guide/api-conventions.md:434]:
```suggestion
API. Nevertheless, migration from annotations to formal schema usually never happens as it requires breaking changes
```

File: dev-guide/api-conventions.md [hunk 2]
```diff
@@ -428,6 +428,24 @@ The simplest line is "no functions".
...
+in customer deployments.
+
+1. Validation does not always come with definition. User set values can be too broad and hard to limit later on.
+2. Lack of discoverability. There's no pre-existing schema that can be published.
+3. Validation is limited. Certain kinds of validators aren't allowed on annotations, so hooks are more frequently used instead.
+4. Hard to extend. An annotation value (a string) can not be extended with additional fields under a parent.
+5. Unclear versioning. Annotation keys can omit versioning. Or, there are multiple ways to specify a version.
+
```
Comment (deads2k) [inline dev-guide/api-conventions.md:442]:
```suggestion
6. Users can "squat" on annotations by adding an unvalidated annotation value for a key that is used in a future version.
```

File: dev-guide/api-conventions.md (hunk 1 above)
Comment (ingvagabund) [reply dev-guide/api-conventions.md:434]: Done

File: dev-guide/api-conventions.md (hunk 2 above)
Comment (ingvagabund) [reply dev-guide/api-conventions.md:442]: Added

**Pull request merged**